address_to_census_tract.py

```diff
@@ -4,84 +4,194 @@
 associated census tract within Washington.
 
 To run:
-    ./address_to_census_tract.py
+    ./address_to_census_tract.py {filepath}
+
+Examples:
+    ./address_to_census_tract.py data/test/testset.json
+    ./address_to_census_tract.py data/test/testset.csv 
 
 Requirements:
-    geocoder
     shapely
+    pandas
+    smartystreets_python_sdk
 """
-import geocoder
 import json
+import os
+from sys import argv
+from textwrap import dedent
 from shapely.geometry import shape, Point
-from sys import stderr
+import logging
+import pandas as pd
+from smartystreets_python_sdk import StaticCredentials, exceptions, ClientBuilder
+from smartystreets_python_sdk.us_street import Lookup
+from smartystreets_python_sdk.us_extract import Lookup as ExtractLookup
+
+LOG = logging.getLogger(__name__)
 
 
 def main():
-    with open("data/test/testset.json", encoding = "UTF-8") as file:
+    if argv[1].endswith('.json'):
+        process_json(argv[1])
+    elif argv[1].endswith('.csv'):
+        process_csv(argv[1])
+    else:
+        raise ValueError(dedent(f"""
+        Unknown file extension for file named «{argv[1]}». 
+        Please choose from one of the following file extensions:
+            * .csv 
+            * .json
+    """))
+
+def process_json(file_path: str):
+    with open(file_path, encoding = "UTF-8") as file:
         testset = [ json.loads(line) for line in file ]
 
     tracts = load_geojson("data/geojsons/Washington_2016.geojson")
 
     for record in testset:
         address = record["address"]
 
+        standard_address = None
+        latlng = None
         tract = None
-        latlng = address_to_latlng(address)
+        response = lookup_address(address)
 
-        if latlng:
+        if response and response['lat'] and response['lng']:
+            standard_address = response['standard_address']
+            latlng = [response['lat'], response['lng']]
             tract = latlng_to_polygon(latlng, tracts)
 
-            if not tract:
-                print(f"failed to find tract for {latlng}", file = stderr)
         else:
-            print(f"failed to geocode {address}", file = stderr)
+            LOG.warning(f"Failed to geocode {address}")
 
-        # GEOID is the nationally-unique tract identifier
         result = {
+            "address": standard_address,
             "latlng": latlng,
-            "tract": tract.get("properties", {}).get("GEOID") if tract else None,
+            "tract": tract,
         }
 
         print(json.dumps({ **record, "result": result }))
 
+def process_csv(file_path: str):
+    """
+    Given a *file_path* to a CSV, processes the `address` column and adds 
+    extra columns for a standardized address, latitude and longitude
+    coordinates, and census tract. If a given address is invalid, these new 
+    columns are left blank. Saves the new table at `data/test/out.csv`. 
```

Comment:
How about outputting the CSV to stdout instead of a hardcoded test file name? (Happy to explain why I think this preferred.)

Reply:
I understand why a hardcoded test file name is not preferred, but why to stdout? Would you rather save the data via
`python address_to_census_tract.py {input_file} > {output_file}`
?

Reply:
The convention of Unix command-line tools is to take input on stdin or named files and send output to stdout or named files.

I suggested stdout because the JSON output is already sent to stdout, and shell redirection is an easy way to save the output. Adding an `--output / -o` option would also be good to name the output file explicitly, but that doesn't have to happen in this PR unless you want to.